### PR TITLE
Cleanup and add category doc

### PR DESCRIPTION
**Pull request by fanzhangio**
**What this PR does / why we need it**:
CategoryExpender interface and some implements structs in package pkg/kubectl/categories are quite confusing to contributors due to the insufficiency documents and illustration.

This PR does:
1. Clean up some useless or vague comments
2. Add illustration for CategoryExpender and implements structs

@pwittrock @droot

### CODE REVIEW

**Comment by fanzhangio** [comment]
Have already signed CLA and config email address of git in the updated commit 7b839f3

File: pkg/kubectl/categories/categories.go [hunk 1]
```diff
@@ -129,8 +146,12 @@ func (e discoveryFilteredExpander) Expand(category string) ([]schema.GroupResour
 	return available, ok
 }
 
+// UnionCategoryExpander represents for a list of CategoryExpander
```
Comment (sttts) [inline pkg/kubectl/categories/categories.go]: and what does it do? returns the union of results?

File: pkg/kubectl/categories/categories.go [hunk 2]
```diff
@@ -21,19 +21,31 @@ import (
 	"k8s.io/client-go/discovery"
 )
 
+// CategoryExpander interface provides a higher layer of group resource mapping.
```
Comment (sttts) [inline pkg/kubectl/categories/categories.go]: the point is that it's an hard-coded category mapping.

File: pkg/kubectl/categories/categories.go [hunk 3]
```diff
@@ -21,19 +21,31 @@ import (
...
+// category can be seen as a classification or 'tag' in the form of key string,
+// which can be used to retrieve a list of groupResource objects
+// 'schema.GroupResource' struct can be seen as a lower layer of group-resource mapping.
 type CategoryExpander interface {
 	Expand(category string) ([]schema.GroupResource, bool)
 }
 
+// SimpleCategoryExpander implements CategoryExpander interface.
```
Comment (sttts) [inline pkg/kubectl/categories/categories.go]: Just say: ... expands based on a hard-coded category list.

File: pkg/kubectl/categories/categories.go [hunk 4]
```diff
@@ -86,14 +99,18 @@ func (e discoveryCategoryExpander) Expand(category string) ([]schema.GroupResour
 	return ret, ok
 }
 
+// discoveryFilteredExpander provides a reverse way comparing to discoveryCategoryExpander.
+// It expends the given CategoryExpander (delegate) which can be seen as a filter, to groupResouce list first (Like a 'target list'),
+// and then request the APIResourceList from server, convert it to groupResouce list based on the server (Like a 'actual list'),
+// lastly compare the two lists, keeping the same elements in result. It performs like a 'Filter'
 type discoveryFilteredExpander struct {
```
Comment (sttts) [inline pkg/kubectl/categories/categories.go]: better try to be shorter. A four line comment like that is hard to maintain and read. What does it do in one sentence?

**Comment by fanzhangio** [comment]
@sttts @droot 
Is it ok to merge ?

File: pkg/kubectl/categories/categories.go [hunk 5]
```diff
@@ -21,10 +21,15 @@ import (
 	"k8s.io/client-go/discovery"
 )
 
+// CategoryExpander interface provides a higher layer of group resource mapping.
```
Comment (sttts) [inline pkg/kubectl/categories/categories.go]: Maybe even shorter: CategoryExpander maps category strings to GroupResouces. Categories are classification or 'tag' of a group of resources.

File: pkg/kubectl/categories/categories.go [hunk 6]
```diff
@@ -62,14 +70,15 @@ func (e discoveryCategoryExpander) Expand(category string) ([]schema.GroupResour
 		if err != nil {
 			return e.fallbackExpander.Expand(category)
 		}
-
+		// Traverse all Category, creating GroupResource object by Group and Resource name, and mapped by category sting as key.
```
Comment (sttts) [inline pkg/kubectl/categories/categories.go]: try to be more abstract and short, maybe something like: collect GroupVersions by categories

File: pkg/kubectl/categories/categories.go [hunk 7]
```diff
@@ -21,10 +21,15 @@ import (
...
 
+// CategoryExpander interface provides a higher layer of group resource mapping.
+// category can be seen as a classification or 'tag' in the form of key string,
+// which can be used to retrieve a list of groupResource objects
+// 'schema.GroupResource' struct can be seen as a lower layer of group-resource mapping.
 type CategoryExpander interface {
 	Expand(category string) ([]schema.GroupResource, bool)
 }
 
+// SimpleCategoryExpander expands based on a hard-coded category list.
```
Comment (droot) [inline pkg/kubectl/categories/categories.go]: SimpleCategoryExpander implements CategoryExpander interface using a static mapping of categories to GroupResource mapping.

File: pkg/kubectl/categories/categories.go [hunk 8]
```diff
@@ -62,14 +70,15 @@ func (e discoveryCategoryExpander) Expand(category string) ([]schema.GroupResour
...
-
+		// Traverse all Category, creating GroupResource object by Group and Resource name, and mapped by category sting as key.
 		for _, apiResource := range apiResourceList.APIResources {
 			if categories := apiResource.Categories; len(categories) > 0 {
 				for _, category := range categories {
 					groupResource := schema.GroupResource{
 						Group:    gv.Group,
 						Resource: apiResource.Name,
 					}
+					// Based on the server returned api resources, fill up discoveredExpansions map
```
Comment (droot) [inline pkg/kubectl/categories/categories.go]: Code below is self-explanatory, so probably, we don't this comment ?

File: pkg/kubectl/categories/categories.go [hunk 9]
```diff
@@ -129,12 +140,16 @@ func (e discoveryFilteredExpander) Expand(category string) ([]schema.GroupResour
 	return available, ok
 }
 
+// UnionCategoryExpander represents for a list of CategoryExpander,
+// expends to a union of gourpResources lists, which merges all expentions of CategoryExpander in UnionCategoryExpander.
 type UnionCategoryExpander []CategoryExpander
 
+//Expand returns a single list of all GroupResource objects. If not found, return empty slice and false
```
Comment (droot) [inline pkg/kubectl/categories/categories.go]: This comment appears to be repeating information. I guess if we say that UnionCategoryExpander implements CategoryExpander interface, then we can skip documenting this method.

File: pkg/kubectl/categories/categories.go [hunk 10]
```diff
@@ -129,12 +140,16 @@ func (e discoveryFilteredExpander) Expand(category string) ([]schema.GroupResour
 	return available, ok
 }
 
+// UnionCategoryExpander represents for a list of CategoryExpander,
```
Comment (droot) [inline pkg/kubectl/categories/categories.go]: UnionCategoryExpander implements CategoryExpander interface. It maps given category string to union of expansions returned by all the CategoryExpanders in the list.

File: pkg/kubectl/categories/categories.go [hunk 11]
```diff
@@ -129,12 +140,16 @@ func (e discoveryFilteredExpander) Expand(category string) ([]schema.GroupResour
...
+// expends to a union of gourpResources lists, which merges all expentions of CategoryExpander in UnionCategoryExpander.
 type UnionCategoryExpander []CategoryExpander
 
+//Expand returns a single list of all GroupResource objects. If not found, return empty slice and false
 func (u UnionCategoryExpander) Expand(category string) ([]schema.GroupResource, bool) {
 	ret := []schema.GroupResource{}
 	ok := false
 
+	// Traverse the UnionCategoryExpander, expend for each CategoryExpander,
```
Comment (droot) [inline pkg/kubectl/categories/categories.go]: may be say: expand the category for each CategoryExpander in the list and merge/combine the results.

File: pkg/kubectl/categories/categories.go [hunk 12]
```diff
@@ -102,6 +112,7 @@ func NewDiscoveryFilteredExpander(delegate CategoryExpander, client discovery.Di
 }
 
 func (e discoveryFilteredExpander) Expand(category string) ([]schema.GroupResource, bool) {
+	// Expends the given CategoryExpander delegate to groupResouce list (as a 'filter')
```
Comment (droot) [inline pkg/kubectl/categories/categories.go]: Code here is self-explanatory, so we can avoid this comment probably ?

File: pkg/kubectl/categories/categories.go (hunk 11 above)
Comment (fanzhangio) [reply pkg/kubectl/categories/categories.go]: Done

File: pkg/kubectl/categories/categories.go (hunk 9 above)
Comment (fanzhangio) [reply pkg/kubectl/categories/categories.go]: done

**Review by fanzhangio — COMMENTED** [review]
Done and updated the PR

**Review by soltysh — APPROVED** [review]
/lgtm

File: pkg/kubectl/categories/categories.go [hunk 13]
```diff
@@ -86,14 +93,15 @@ func (e discoveryCategoryExpander) Expand(category string) ([]schema.GroupResour
 	return ret, ok
 }
 
+// discoveryFilteredExpander expands the given CategoryExpander (delegate) to filter group and resrouce returned from server
```
Comment (sttts) [inline pkg/kubectl/categories/categories.go]: typo: resource

Comment (fanzhangio) [reply pkg/kubectl/categories/categories.go]: Done! Thanks

**Comment by fanzhangio** [comment]
@sttts Could you plz give me a lgtm label for this PR ? Thank you so much

**Comment by droot** [comment]
/assign @janetkuo 
Hey Janet, can you pl. approve this PR. Its been lgtm already.

**Review by janetkuo — APPROVED** [review]

**Pull request merged**